Updating pycoreaudio to python3

Hi @larsimmisch 

I've updated the pycoreaudio extension to python3 c-api in my [fork](https://github.com/shakfu/pycoreaudio). My original intent was to create a cython wrapper, but I started fixing the original extension to the extent I could.

There are still a  bunch of deprecations, and it certainly needs more testing and polish but ... would you be interested in a PR and to update this extension to 3 properly? It would be a shame not to return this nice project to usefulness to python3 users...

Thanks for your reply... I will make a PR soonish... fyi, the current deprecations / warnings are:

```c++
coreaudio.c:339:9: warning: 'CloseComponent' is deprecated: first deprecated in macOS 10.8 [-Wdeprecated-declarations]
        CloseComponent(obj->instance);
        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/System/Library/Frameworks/CoreServices.framework/Frameworks/CarbonCore.framework/Headers/Components.h:594:1: note: 'CloseComponent' has been explicitly marked deprecated here
CloseComponent(ComponentInstance aComponentInstance)          __OSX_AVAILABLE_BUT_DEPRECATED(__MAC_10_0, __MAC_10_8, __IPHONE_NA, __IPHONE_NA);
^
coreaudio.c:636:13: warning: 'FindNextComponent' is deprecated: first deprecated in macOS 10.8 [-Wdeprecated-declarations]
        c = FindNextComponent(NULL, &componentDescription->desc);
            ^
/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/System/Library/Frameworks/CoreServices.framework/Frameworks/CarbonCore.framework/Headers/Components.h:470:1: note: 'FindNextComponent' has been explicitly marked deprecated here
FindNextComponent(
^
coreaudio.c:638:13: warning: 'FindNextComponent' is deprecated: first deprecated in macOS 10.8 [-Wdeprecated-declarations]
        c = FindNextComponent(component->component,
            ^
/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/System/Library/Frameworks/CoreServices.framework/Frameworks/CarbonCore.framework/Headers/Components.h:470:1: note: 'FindNextComponent' has been explicitly marked deprecated here
FindNextComponent(
^
coreaudio.c:665:10: warning: 'OpenAComponent' is deprecated: first deprecated in macOS 10.8 [-Wdeprecated-declarations]
    rc = OpenAComponent(component->component, &au);
         ^
/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/System/Library/Frameworks/CoreServices.framework/Frameworks/CarbonCore.framework/Headers/Components.h:560:1: note: 'OpenAComponent' has been explicitly marked deprecated here
OpenAComponent(
^
coreaudio.c:708:5: warning: 'PyEval_InitThreads' is deprecated [-Wdeprecated-declarations]
    PyEval_InitThreads();
    ^
/opt/homebrew/opt/python@3.12/Frameworks/Python.framework/Versions/3.12/include/python3.12/ceval.h:132:1: note: 'PyEval_InitThreads' has been explicitly marked deprecated here
Py_DEPRECATED(3.9) PyAPI_FUNC(void) PyEval_InitThreads(void);
^
/opt/homebrew/opt/python@3.12/Frameworks/Python.framework/Versions/3.12/include/python3.12/pyport.h:317:54: note: expanded from macro 'Py_DEPRECATED'
#define Py_DEPRECATED(VERSION_UNUSED) __attribute__((__deprecated__))
                                                     ^
5 warnings generated.
```

--- a/coreaudio.c
+++ b/coreaudio.c
@@ -399,7 +399,7 @@ static OSStatus audio_unit_render_callback(
     if (o != Py_None) {
         if (!PyLong_Check(o)) {
             fprintf(stderr,
-                    "render callback must return a tuple (int, buffers)\n");
+                    "render callback must return a tuple (None|int, bytes)\n");
             goto error;
         }
 
@@ -411,9 +411,9 @@ static OSStatus audio_unit_render_callback(
         Py_ssize_t len;
 
         o = PyTuple_GetItem(result, i);
-        if (!PyUnicode_Check(o)) {
+        if (!PyBytes_Check(o)) {
             fprintf(stderr,
-                    "render callback must return a tuple (int, buffers)\n");
+                    "render callback must return a tuple (None|int, bytes)\n");
             goto error;
         }
         if (PyBytes_AsStringAndSize(o, &buffer, &len) < 0)
@@ -701,11 +701,13 @@ static PyModuleDef coreaudiomodule = {
 
 PyMODINIT_FUNC
 PyInit_coreaudio(void)
-// void initcoreaudio(void)
 {
-    PyObject* m;
 
-    // PyEval_InitThreads();
+#if PY_VERSION_HEX < 0x03090000
+    PyEval_InitThreads();
+#else
+    Py_Initialize();
+#endif
 
     if (PyType_Ready(&AudioComponentDescriptionType) < 0)
         return NULL;
@@ -722,7 +724,7 @@ PyInit_coreaudio(void)
     if (PyType_Ready(&AudioUnitType) < 0)
         return NULL;
 
-    m = PyModule_Create(&coreaudiomodule);
+    PyObject* m = PyModule_Create(&coreaudiomodule);
     if (m == NULL)
         return NULL;
 

--- a/play.py
+++ b/play.py
@@ -1,7 +1,6 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
 
 import coreaudio
-import time
 import threading
 from optparse import OptionParser
 import wave
@@ -17,7 +16,7 @@ def fourcctoi(v):
     return (ord(v[0]) << 24) + (ord(v[1]) << 16) + (ord(v[2]) << 8) + ord(v[3])
 
 def au_wav_prepare(au, fn, verbose = False):
-    """Open a wav file called 'fn' and set the stream format based upon the 
+    """Open a wav file called 'fn' and set the stream format based upon the
     information in the wav header.
 
     Returns the open file."""
@@ -30,7 +29,7 @@ def au_wav_prepare(au, fn, verbose = False):
     channels: %d
     sample width: %d
     """ % (fn, f.getframerate(), f.getnchannels(), f.getsampwidth()))
-               
+
     sd = coreaudio.AudioStreamBasicDescription(
         f.getframerate(),
         coreaudio.kAudioFormatLinearPCM,
@@ -50,8 +49,8 @@ def play(au, f):
     def render_callback(flags, time, bus, frames, nbuffers, user_data):
         done.acquire()
         buf = f.readframes(frames)
-        # print 'read %d bytes' % len(buf)
-        
+        # print(f'read {len(buf)} bytes')
+
         if not buf:
             done.notify()
             done.release()
@@ -65,7 +64,7 @@ def play(au, f):
         if sil:
             buf = buf + '\0' * sil
         return (None, buf)
-    
+
     done = threading.Condition()
     done.acquire()
 
@@ -73,7 +72,7 @@ def play(au, f):
     au.SetRenderCallback(render_callback)
     print('Starting')
     au.Start()
-    
+
     done.wait()
     au.SetRenderCallback(None)
 
@@ -107,7 +106,7 @@ if __name__ == '__main__':
     options, args = parser.parse_args()
     if len(args) < 1:
         parser.error('need at least one file argument')
-    
+
     au = open_default_au(options.manufacturer)
 
     for a in args:

